### PR TITLE
ruby_3_2 backport for #8006

### PR DESCRIPTION
**Pull request by XrXr**
This is a backported version of https://github.com/ruby/ruby/pull/8006
The patch is different because the two branches have diverged.

Original bug report: https://github.com/Shopify/yjit/issues/313

---

YJIT: Fix autosplat miscomp for blocks with optionals

When passing an array as the sole argument to `yield`, and the yieldee takes more than 1 optional parameter, the array is expanded similar to `*array` splat calls. This is called "autosplat" in `setup_parameters_complex()`.

Previously, YJIT did not detect this autosplat condition. It passed the array without expanding it, deviating from interpreter behavior. Detect this conditon and refuse to compile it.

### CODE REVIEW

**Review by k0kubun — APPROVED** [review]

**Comment by maximecb** [comment]
Everything passed except for something which seems to be a Cirrus CI issue, so I'm going to merge this.

**Comment by XrXr** [comment]
Sorry, I forgot to mention that we're not supposed to merge changes to backport branches (though it's fine to make requests). The branch maintainer is given complete control over what lands. At least that's the policy last I checked.

**Comment by hsbt** [comment]
@maximecb Don't merge without @nagachika approval at `ruby_3_2` branch.

**Comment by nagachika** [comment]
I have reviewed the changeset.

Also, if the changesets in the master branch cannot be cleanly applied, creating patches for the stable branches is always welcome. I will handle PRs when I am mentioned or assigned.

**Pull request merged**
